net/libfab/libfab.c

```diff
@@ -327,15 +347,89 @@ static int libfab_buf_register(struct m0_net_buffer *nb)
  */
 static int libfab_buf_add(struct m0_net_buffer *nb)
 {
-	int        result = 0;
-	/*
- 	* TODO:
- 	*   fi_send/fi_recv
- 	* */
-	return M0_RC(result);
-}
-
+	struct m0_fab__buf_params	*fbp;
+	struct m0_fab__tm		*ma;
+	struct fid_mr			*mr;
+	struct m0_fab__ep		*ep;
+	struct m0_fab__bdesc		*peer;
+	fi_addr_t 			 remote_rx_addr;
+	int			  	 qt;
+	int				 ret = 0;
+	ssize_t				 cnt = 0;
+
+	M0_PRE(m0_mutex_is_locked(&nb->nb_tm->ntm_mutex));
```

Comment:
Instead of multiple check can we add 
```suggestion
	M0_PRE(m0_mutex_is_locked(&nb->nb_tm->ntm_mutex));
```
	struct m0_fab__buf *buf = nb->nb_xprt_private;
	struct m0_fab__tm  *ma  = libfab_buf_tm(buf);
	
	M0_PRE(libfab_tm_is_locked(ma) && libfab_tm_invariant(ma) &&
	       libfab_buf_invariant(buf));`